From b29dc466f0c06e81868b7bfa780019c309904310 Mon Sep 17 00:00:00 2001
From: Emergi Idan <idan.emergi@intel.com>
Date: Tue, 17 Mar 2020 15:11:39 +0200
Subject: [PATCH] PONRTSYS-7025: ppv4 qos potential wrr inaccuracy when setting
 more than 8 queues

Print WRR weights in full tree view and make sure
the actual configured weight will exceed the maximum
value supported by the HW

Signed-off-by: Emergi Idan <idan.emergi@intel.com>
---
 drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c    |  2 ++
 drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_utils.c | 35 ++++++++++++++++-------
 2 files changed, 26 insertions(+), 11 deletions(-)

diff --git a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
index d683b3cf3e71..5fd4ad7f3271 100644
--- a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
+++ b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_fw.c
@@ -409,6 +409,7 @@ static const char *const cmd_str[] = {
 	[CMD_TYPE_SET_QUEUE] = "CMD_TYPE_SET_QUEUE",
 	[CMD_TYPE_REM_QUEUE] = "CMD_TYPE_REM_QUEUE",
 	[CMD_TYPE_GET_QUEUE_STATS] = "CMD_TYPE_GET_QUEUE_STATS",
+	[CMD_TYPE_GET_QM_STATS] = "CMD_TYPE_GET_QM_STATS",
 	[CMD_TYPE_SET_AQM_SF] = "CMD_TYPE_SET_AQM_SF",
 	[CMD_TYPE_CLK_UPDATE] = "CMD_TYPE_CLK_UPDATE",
 	[CMD_TYPE_NODES_REFRESH] = "CMD_TYPE_NODES_REFRESH",
@@ -426,6 +427,7 @@ static const char *const cmd_str[] = {
 	[CMD_TYPE_FLUSH_QUEUE] = "CMD_TYPE_FLUSH_QUEUE",
 	[CMD_TYPE_GET_SYSTEM_INFO] = "CMD_TYPE_GET_SYSTEM_INFO",
 	[CMD_TYPE_UPDATE_PORT_TREE] = "CMD_TYPE_UPDATE_PORT_TREE",
+	[CMD_TYPE_SUSPEND_PORT_TREE] = "CMD_TYPE_SUSPEND_PORT_TREE",
 	[CMD_TYPE_GET_QUANTA] = "CMD_TYPE_GET_QUANTA",
 	[CMD_TYPE_INTERNAL] = "CMD_TYPE_INTERNAL"
 };
diff --git a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_utils.c b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_utils.c
index a62e61befc70..0804bcad86a0 100644
--- a/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_utils.c
+++ b/drivers/net/ppv4/pp-api/pp-qos-mgr/pp_qos_utils.c
@@ -1865,6 +1865,8 @@ static void set_children_bw_weight(struct pp_qos_dev *qdev,
 	for (; cnt; --cnt) {
 		QOS_ASSERT(node_child(child), "Node is not a child\n");
 		bw_weight = child->child_prop.virt_bw_share / gcd;
+		/* make sure bw_weight will not exceed the maximum */
+		bw_weight = min_t(u32, bw_weight, QOS_MAX_CHILD_WRR_WEIGHT);
 		child->child_prop.bw_weight = bw_weight;
 		send_bw_weight_command(qdev, child, bw_weight);
 		++child;
@@ -3961,11 +3963,12 @@ static void qos_dbg_dump_subtree(struct pp_qos_dev *qdev, struct qos_node *node,
 	char tabs_str[64];
 	char indent_str[64];
 	char aqm_codel_str[64];
+	char weight_str[64] = "";
+	char bw_str[64] = "";
 	bool last_child;
 	bool aqm_queue;
 	bool codel_queue;
 	struct qos_node *child;
-	char bw_str[64];
 	u32 max_bw = qdev->hwconf.max_bw_limit_kbps;
 
 	if (unlikely(depth > QOS_MAX_PREDECESSORS)) {
@@ -4000,14 +4003,19 @@ static void qos_dbg_dump_subtree(struct pp_qos_dev *qdev, struct qos_node *node,
 		if (node_sched(child)) {
 			if (child->bandwidth_limit &&
 			    child->bandwidth_limit != max_bw)
-				snprintf(bw_str, 64, ", bw %u)",
+				snprintf(bw_str, 64, ", bw %u",
 					 child->bandwidth_limit);
-			else
-				snprintf(bw_str, 64, ")");
 
-			QOS_DBG_PRINT(s, "%sSched%s-%u(%u%s-%s\n",
+			if (!user_tree && (node->parent_prop.arbitration ==
+			    PP_QOS_ARBITRATION_WRR))
+				snprintf(weight_str, sizeof(weight_str),
+					 ", w %u, vw %u",
+					 child->child_prop.bw_weight,
+					 child->child_prop.virt_bw_share);
+
+			QOS_DBG_PRINT(s, "%sSched%s-%u(%u%s%s)-%s\n",
 				      indent_str, node_rep(child) ? "'" : "",
-				      node_id, child_phy, bw_str,
+				      node_id, child_phy, bw_str, weight_str,
 				      ARB_STR(child->parent_prop.arbitration));
 			qos_dbg_dump_subtree(qdev, child, depth + 1, depth + 1,
 					     user_tree, s);
@@ -4026,18 +4034,23 @@ static void qos_dbg_dump_subtree(struct pp_qos_dev *qdev, struct qos_node *node,
 			if ((child->bandwidth_limit &&
 			     child->bandwidth_limit != max_bw) ||
 			    child->data.queue.eir)
-				snprintf(bw_str, 64, ", bw %u/%u)",
+				snprintf(bw_str, 64, ", bw %u/%u",
 					 child->bandwidth_limit,
 					 child->data.queue.eir);
-			else
-				snprintf(bw_str, 64, ")");
 
-			QOS_DBG_PRINT(s, "%s%sQueue%s-%u(%u%s-rlm-%u%s\n",
+			if (!user_tree && (node->parent_prop.arbitration ==
+			    PP_QOS_ARBITRATION_WRR))
+				snprintf(weight_str, sizeof(weight_str),
+					 ", w %u, vw %u",
+					 child->child_prop.bw_weight,
+					 child->child_prop.virt_bw_share);
+
+			QOS_DBG_PRINT(s, "%s%sQueue%s-%u(%u%s%s)-rlm-%u%s\n",
 				      indent_str,
 				      node_syncq(child) ? "Sync-" : "",
 				      node_rep(child) ? "'" : "",
 				      node_id, child_phy,
-				      bw_str,
+				      bw_str, weight_str,
 				      child->data.queue.rlm,
 				      (aqm_queue || codel_queue) ?
 				      aqm_codel_str : "");
